File: mobile/Claw/AppSettingsAffordance.swift
import Foundation

/// The "App settings" affordance, read from `core/spec/app-settings-affordance.json`,
/// plus the shim this client installs so the injected script can reach the app's
/// own settings sheet.
///
/// ONE script, TWO clients. The script itself is not ported: it is read from the
/// spec the app bundles, exactly as `PromptMetadata` reads its hook, so the bytes
/// the desktop installs into the gateway page and the bytes this client installs
/// through a `WKUserScript` are one copy rather than two dialects. What differs is
/// only the host bridge the script calls, `window.__clawAppSettings.open`, and it
/// differs only in HOW open() reaches a window: an IPC on the desktop, a
/// `WKScriptMessageHandler` here.
///
/// The affordance never reimplements settings. It finds the Control UI's sidebar
/// footer, adds a control, and on a click calls the bridge; the surface it opens
/// is the same `core/ui/settings.html` this client already loads in a sheet. See
/// `AppSettingsBridge` for the handler, and `WebView` for where both are installed.
///
/// The spec is read at runtime rather than mirrored as Swift constants, for the
/// same reason `PromptMetadata`'s is: a script cannot be mirrored without a second
/// copy existing, and a second copy is the fork the shared file exists to prevent.
enum AppSettingsAffordance {
    private struct Spec: Decodable {
        /// The handoff's timing, which both clients must hold to the same line.
        /// Read from the spec for the same reason the route is: a number two
        /// clients have to agree on needs one owner. See
        /// `readyTimeoutMs` and `pollMs` in the spec's `handoff` object.
        struct Handoff: Decodable {
            let readyTimeoutMs: Int
            let pollMs: Int
        }

        let global: String
        let configGlobal: String
        let marker: String
        let anchors: [String: String]
        /// The Control UI route the script may hand the reader to when there is no
        /// control to press, read from the same spec as the anchors. Decoded by
        /// NAME rather than mirrored in Swift, because it is a path the Control UI
        /// owns and a second copy could be wrong the day upstream moves a route.
        let routes: [String: String]?
        let handoff: Handoff
        let script: [String]
    }

    private static let spec: Spec = loadSpec()

    private static func loadSpec() -> Spec {
        let empty = Spec(global: "", configGlobal: "", marker: "", anchors: [:], routes: nil,
                         handoff: Spec.Handoff(readyTimeoutMs: 0, pollMs: 0), script: [])
        guard let url = Bundle.main.url(forResource: "app-settings-affordance", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let spec = try? JSONDecoder().decode(Spec.self, from: data),
              !spec.global.isEmpty,
              !spec.script.isEmpty
        else {
            // A build that did not bundle the spec cannot inject the affordance.
            // It installs nothing rather than something invented, and
            // `AppSettingsAffordanceParityTests` is what turns that into a failing
            // build rather than a quiet absence in the field.
            return empty
        }
        return spec
    }

    /// The global the client installs its host bridge on. Read by the script,
    /// never written by it.
    static var global: String { spec.global }

    /// The plain, writable global the config (label, tooltip, tokens, anchors)
    /// is set on. Separate from the bridge global so the script never assigns to
    /// a frozen object.
    static var configGlobal: String { spec.configGlobal }

    /// The attribute the injected control carries, so a re-render cannot stack two.
    static var marker: String { spec.marker }

    /// The injected script, exactly as the spec holds it. The desktop installs
    /// these same bytes over `executeJavaScript`; this client installs them
    /// through a `WKUserScript`.
    static var script: String { spec.script.joined(separator: "\n") }

    /// The label and tooltip the control reads, plus the resolved tokens it
    /// styles itself with. The tokens are left empty here: the script falls back
    /// to the page's own `currentColor` and neutral values, so the control reads
    /// as part of the footer without this client having to resolve the Control
    /// UI's palette, which is the page's to own.
    private static var config: [String: Any] {
        [
            "label": "App settings",
            "tooltip": "\(Naming.product) settings",
            "anchors": spec.anchors,
            // The same two things the desktop's configStatement() hands over, from
            // the same spec: the anchors, and the Control UI's own route for the
            // destination its settings entry opens.
            "routes": spec.routes ?? [String: String](),
            "tokens": [String: String](),
        ]
    }

    /// The call that takes the reader to the CONTROL UI's own settings.
    ///
    /// The port of `controlUiSettingsSource()` in
    /// `core/app-settings-affordance.js`, and the same split as everything else
    /// here: the script that does the pressing is the shared one (bundled and
    /// installed above), and this is the one line that asks the page to run it.
    /// The function is left on the CONFIG global by the installation, which is
    /// why this looks it up there rather than on the bridge: the bridge may be a
    /// frozen object on another client, and the config global is a plain one.
    static func controlUiSettingsSource() -> String {
        """
        (function () {
          try {
            var config = window.\(spec.configGlobal);
            if (config && typeof config.openControlUiSettings === 'function') return config.openControlUiSettings();
          } catch (e) { return false; }
          return false;
        })()
        """
    }

    /// How long to hold our own surface waiting for the destination, and how
    /// often to ask. The spec's, so the phone and the desktop wait alike.
    static var readyTimeoutMs: Int { spec.handoff.readyTimeoutMs }
    static var pollMs: Int { spec.handoff.pollMs }

    /// The question that tells this client whether the destination has ARRIVED.
    ///
    /// The port of `controlUiSettingsReadySource()` in
    /// `core/app-settings-affordance.js`, built from the same spec fields, so the
    /// claim about the Control UI's DOM has one owner rather than one per client.
    /// Asked repeatedly against the live page, never awaited inside it: the
    /// shipping Control UI has no footer settings control to press, so the ask
    /// falls through to its own settings ROUTE, and a route is a full document
    /// load that destroys any promise waiting on it.
    ///
    /// Both halves are necessary and neither is sufficient. The route is what
    /// makes it the page the reader asked for, and also what makes the answer mean
    /// "arrived" rather than "already there": the settings shell is up on every
    /// `/settings/*` route, including the first-run flow, so a reader who came
    /// from the Control UI's own settings page would satisfy a node-only question
    /// the instant they pressed. The painted node is what makes it rendered rather
    /// than merely committed.
    ///
    /// `nil` when this build's spec carries no surface or route to look for, which
    /// is a broken bundle rather than a state to guess at: the caller skips the
    /// wait and says so, rather than holding the reader against a question it
    /// cannot ask. A shipped build cannot reach this, because
    /// `AppSettingsAffordanceParityTests` asserts the spec carries both.
    static func controlUiSettingsReadySource() -> String? {
        guard let surface = spec.anchors["controlUiSettingsSurface"], !surface.isEmpty,
              let route = spec.routes?["appearance"], !route.isEmpty
        else { return nil }
        return """
        (function () {
          try {
            if (location.pathname !== \(jsonString(route))) return false;
            var node = document.querySelector(\(jsonString(surface)));
            if (!node) return false;
            var box = node.getBoundingClientRect();
            return box.width > 0 && box.height > 0;
          } catch (e) { return false; }
        })()
        """
    }

    /// A Swift string as a JavaScript string literal. The values spliced into the
    /// statements above are the Control UI's own selectors and route, so a quote
    /// or a backslash in one is unlikely rather than impossible, and an escaping
    /// mistake would produce a script that throws inside a page we do not own.
    private static func jsonString(_ value: String) -> String {
        guard let data = try? JSONEncoder().encode(value),
              let json = String(data: data, encoding: .utf8)
        else { return "\"\"" }
        return json
    }

    /// The bridge shim, then the configuration, then the shared script.
    ///
    /// The bridge is installed on its own global; the config on a separate plain
    /// global; the script reads both. Posting through
    /// `window.webkit.messageHandlers` is the one route a page has back into this
    /// app, the same channel the theme relay and the settings host use; `open`
    /// takes no argument, so there is nothing to carry but the ask.
    ///
    /// Unlike the desktop's contextBridge object, this bridge global is a plain
    /// object the shim assigns, so the two globals could in principle be one here;
    /// they are kept separate so the phone and the desktop install the SAME shared
    /// script, which reads a frozen bridge on the desktop and must never write it.
    static func installation() -> String {
        let configJSON: String = {
            guard let data = try? JSONSerialization.data(withJSONObject: config, options: [.sortedKeys]),
                  let json = String(data: data, encoding: .utf8)
            else { return "{}" }
            return json
        }()
        let bridge = """
        window.\(spec.global) = {
          open: function () {
            try { window.webkit.messageHandlers.\(AppSettingsBridge.messageName).postMessage({}); } catch (e) {}
          }
        };
        """
        return """
        \(bridge)
        window.\(spec.configGlobal) = \(configJSON);
        \(script)
        """
    }
}

File: core/app-settings-affordance.js
// The affordance that adds an "App settings" control to the Control UI's
// sidebar footer, and the host bridge it calls to open the app's own settings
// surface.
//
// ONE script, TWO clients. This is the platform-free half: the script itself
// lives in spec/app-settings-affordance.json so the bytes the desktop runs and
// the bytes the phone run are one copy rather than two that agree by
// convention, exactly as the client-context hook is (see prompt-metadata.js).
// The desktop reads this module and installs the script into the gateway page
// over executeJavaScript; the iOS client bundles the spec and installs it
// through a WKUserScript, reading the same fields this module exposes.
//
// What the affordance does NOT do is reimplement settings. It finds the footer,
// adds a control, and on a click calls a host bridge the client wires up. That
// bridge is the one thing that differs between clients, and it differs only in
// how open() reaches the app's window: an IPC on the desktop, a message handler
// on the phone. The shared settings surface it opens is core/ui/settings.html,
// which both clients already load; this only opens it.
//
// Consumers: desktop/src/main.js installs installation() into the gateway page
// and answers the bridge in desktop/src/preload.cjs; iOS installs the same
// script through a WKUserScript and answers the bridge in a message handler.

import spec from './spec/app-settings-affordance.json' with { type: 'json' };

/** The global the client installs its host bridge on. May be a FROZEN contextBridge object, so the script only reads it. */
export const AFFORDANCE_GLOBAL = spec.global;

/** The plain, writable global the config (label, tooltip, tokens, anchors) is set on. */
export const AFFORDANCE_CONFIG_GLOBAL = spec.configGlobal;

/** The attribute the injected control carries, so a re-render cannot stack two of them. */
export const AFFORDANCE_MARKER = spec.marker;

/** The selectors the script tries in order: the footer actions, the footer bar, then any sidebar. */
export const AFFORDANCE_ANCHORS = spec.anchors;

/**
 * The Control UI routes a client may hand the reader to, by name.
 *
 * One today: the route the Control UI's own settings entry opens, which is the
 * destination "Go to the Control UI settings" promises. Preferred mechanism is
 * still to press the Control UI's own control; this is what happens when there
 * is no control to press, so the button cannot silently do nothing.
 */
export const AFFORDANCE_ROUTES = spec.routes;

/** The default label and tooltip, so a client that passes neither still reads sensibly. */
export const DEFAULT_LABEL = 'App settings';
export const DEFAULT_TOOLTIP = 'Open this app\u2019s settings';

/**
 * How long a client holds its own surface waiting for the Control UI's settings
 * page to arrive, and how often it asks.
 *
 * Read from the spec rather than written here, because the phone has to hold the
 * same line and a number that two clients must agree on needs one owner. A
 * DEADLINE rather than an open wait, and the reason is the failure it bounds: a
 * Control UI that never reaches the route (a renamed surface node, a gateway that
 * redirects away, a page that is simply broken) must not leave a reader behind a
 * surface that never lets go. Past the deadline the client reveals anyway and logs
 * it, which is the behaviour of the day before rather than a new way to be stuck.
 *
 * The number is generous on purpose. The gap this exists to remove is the
 * destination's own load, which on a remote gateway is seconds rather than
 * milliseconds, and holding a working settings page a moment longer is strictly
 * better than showing a page the reader did not ask for. What it is NOT is
 * unbounded: ten seconds is longer than any gateway this app talks to takes to
 * paint its own settings page, and a surface still holding at that point is
 * reporting a fault rather than waiting on a load.
 */
export const CONTROL_UI_SETTINGS_READY_TIMEOUT_MS = spec.handoff.readyTimeoutMs;

/** The interval between readiness questions. Cheap: the statement is one boolean. */
export const CONTROL_UI_SETTINGS_POLL_MS = spec.handoff.pollMs;

/**
 * The injected script, exactly as the spec holds it. One copy, two engines.
 *
 * The desktop runs these bytes through executeJavaScript and the phone through
 * a WKUserScript; neither ports it, because a port is a second copy and a second
 * copy is the fork this file exists to prevent.
 */
export function affordanceSource() {
  return spec.script.join('\n');
}

/**
 * The configuration the script reads, as the statement that sets it.
 *
 * A separate statement ahead of the script rather than text spliced into it, so
 * the script body has no per-platform parts and both engines run identical
 * bytes. The client supplies the label, the tooltip and the resolved design
 * tokens; the anchors come from the spec so a selector change is one edit.
 *
 * Set on the CONFIG global rather than the bridge global. The desktop's bridge
 * is a frozen contextBridge object, so a reassignment or an Object.assign onto
 * it throws in the page; the config lives on its own plain global instead, and
 * the script only ever reads the bridge. The `open` function is not set here: it
 * is the bridge the client installs, reached at click time, and the two clients
 * reach their own windows in two different ways.
 */
export function configStatement({ label = DEFAULT_LABEL, tooltip = DEFAULT_TOOLTIP, tokens = {} } = {}) {
  const config = {
    label: String(label || DEFAULT_LABEL),
    tooltip: String(tooltip || DEFAULT_TOOLTIP),
    anchors: spec.anchors,
    // The Control UI's own route for the destination its settings entry opens.
    // Handed in like the anchors, so a route change is one edit in the spec
    // rather than one per client, and so the script that uses it runs identical
    // bytes on both.
    routes: spec.routes,
    tokens: tokens && typeof tokens === 'object' ? tokens : {},
  };
  return `window.${spec.configGlobal} = ${JSON.stringify(config)};`;
}

/**
 * The call that takes the reader to the CONTROL UI's own settings.
 *
 * A separate statement from the installation, the same shape as the config
 * statement, because it is evaluated at a different moment by a different
 * caller: installation happens once per page at document start, and this is
 * asked for when someone presses the button on our settings page. The script
 * installation left the function on the CONFIG global (which is writable; the
 * bridge may be frozen), so this only has to look it up and call it.
 *
 * It returns whether a control was there to press, which the caller logs rather
 * than hides: a fail-soft placement is fine for a control nobody pressed, and a
 * fail-soft navigation is a button that appears to do nothing, which is worth a
 * line in the app's own stdout.
 */
export function controlUiSettingsSource() {
  return `(function () {
  try {
    var config = window.${spec.configGlobal};
    if (config && typeof config.openControlUiSettings === 'function') return config.openControlUiSettings();
  } catch (e) { return false; }
  return false;
})()`;
}

/**
 * The question that tells a client whether the destination has ARRIVED.
 *
 * The other half of the handoff, and the reason the handoff is atomic. The
 * source above asks the Control UI to show its settings; this one answers
 * whether it has, so a client can hold its own surface until there is something
 * to reveal rather than dismissing into a load it cannot see the end of.
 *
 * Asked REPEATEDLY against the live page rather than awaited inside the page,
 * and that is forced by the mechanism rather than chosen: the shipping Control UI
 * has no footer settings control, so the ask falls through to the Control UI's own
 * route, and a route is a FULL DOCUMENT LOAD. A promise returned by the page is
 * destroyed by the very navigation it is waiting on, so the wait has to be driven
 * from outside the document, where it can survive the swap and re-ask the new one.
 *
 * Both halves of the answer are necessary and neither is sufficient. The ROUTE is
 * what makes it the destination the reader asked for, and it is also what makes
 * the answer mean "arrived" rather than "already there": the settings shell is up
 * on every `/settings/*` route, so a reader who came from the Control UI's own
 * settings page would satisfy a node-only question at the instant they pressed.
 * The NODE, painted, is what makes it rendered rather than committed, since the
 * route lands a frame or more before the page has drawn anything.
 *
 * The selectors and the route come from the spec, spliced in as JSON, so there is
 * one owner of every Control UI fact this depends on and the Swift client builds
 * the same statement from the same file rather than a hand-copied selector.
 */
export function controlUiSettingsReadySource() {
  const surface = spec.anchors.controlUiSettingsSurface;
  const route = spec.routes.appearance;
  return `(function () {
  try {
    if (location.pathname !== ${JSON.stringify(route)}) return false;
    var node = document.querySelector(${JSON.stringify(surface)});
    if (!node) return false;
    var box = node.getBoundingClientRect();
    return box.width > 0 && box.height > 0;
  } catch (e) { return false; }
})()`;
}
/**
 * What a client installs into the gateway page: the configuration, then the
 * shared script. The client installs its bridge on `window.<AFFORDANCE_GLOBAL>`
 * separately (the desktop in its preload, the phone in a shim); the script reads
 * that bridge at click time rather than at install time, so the order of the two
 * installs does not matter.
 *
 * Re-installing is safe: the configuration is a plain assignment, and the script
 * no-ops if the control is already placed.
 */
export function installation(options = {}) {
  return `${configStatement(options)}\n${affordanceSource()}`;
}
